src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java

```diff
@@ -0,0 +1,181 @@
+package com.epam.esm.dao.impl;
+
+
+
+import com.epam.esm.dao.CertificateDao;
+import com.epam.esm.dao.DaoException;
+import com.epam.esm.model.Certificate;
+import com.epam.esm.model.Tag;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataAccessException;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.jdbc.core.BeanPropertyRowMapper;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.PreparedStatementCreator;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.BeanPropertySqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.MultiValueMap;
+
+import java.sql.*;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.epam.esm.dao.RowMapProvider.*;
+import static com.epam.esm.dao.SqlProvider.*;
+
+@Repository
+@Transactional(readOnly = true)
+public class CertificateDaoImpl implements CertificateDao, RowMapper<Certificate> {
+
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+
+    @Autowired
+    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+    private List<Long> getCertificateIdsByTagId(Long tagId){
+        return jdbcTemplate.query(GET_CERTIFICATES_BY_TAG_ID, new Object[] {tagId},
+                (rs, rowNum) -> rs.getLong(1));
+    }
+
+
+    @Override
+    public List<Certificate> getAll() {
+        return jdbcTemplate.query(GET_ALL_CERTIFICATES, new CertificateDaoImpl());
+
+    }
+
+    @Override
+    public List<Certificate> getByParameters(MultiValueMap<String, String> params) throws DaoException {
```

Comment:
You don't need to throw Exception if you extend RuntimeException.

Reply:
In my method, i want to throw an exception when collection is empty